pkg/controller/machine/machine.go

```diff
@@ -461,16 +462,23 @@ func (c *Controller) ensureMachineHasNodeReadyCondition(machine *clusterv1alpha1
 
 // deleteMachine makes sure that an instance has gone in a series of steps.
 func (c *Controller) deleteMachine(prov cloud.Provider, machine *clusterv1alpha1.Machine) error {
+	if machine.Status.NodeRef != nil {
+		_, err := c.nodesLister.Get(machine.Status.NodeRef.Name)
+		if err != nil {
+			if !kerrors.IsNotFound(err) {
+				return fmt.Errorf("failed to get node %s for machine %s/%s: %v", machine.Status.NodeRef.Name, machine.Namespace, machine.Name, err)
+			}
+		}
+		if err := eviction.New(machine.Status.NodeRef.Name, c.nodesLister, c.kubeClient).Run(); err != nil {
+			return fmt.Errorf("failed to evict node %s: %v", machine.Status.NodeRef.Name, err)
+		}
+	}
+
 	if err := c.deleteCloudProviderInstance(prov, machine); err != nil {
 		c.recorder.Eventf(machine, corev1.EventTypeWarning, "DeletionFailed", "Failed to delete instance at cloud provider: %v", err)
 		return err
 	}
 
-	// Delete the node object after the instance is gone
```

Comment:
This check is already in `deleteNodeForMachine` so having it here was redundant

Reply:
https://github.com/kubermatic/machine-controller/pull/346/files#diff-87acff32b00826238809db10ac9ef562L530

Reply:
Correction: This finalizer here is not in `deleteNodeForMachine` but it gets removed by `deleteCloudProviderInstance` so at this point here it is impossible that it still exists